modules/katamari/src/main/ts/ephox/katamari/api/Optional.ts

```diff
@@ -1,141 +1,290 @@
+    return new Optional(true, value);
+  }
+
+  /**
+   * Create a new `Optional<T>` that **does not** contain a value. `T` can be
+   * any type because we don't actually have a `T`.
+   */
+  public static none<T = never>(this: void): Optional<T> {
+    return Optional.singletonNone;
+  }
+
+  /**
+   * Perform a transform on an `Optional` type. Regardless of whether this
+   * `Optional` contains a value or not, `fold` will return a value of type `U`.
+   * If this `Optional` does not contain a value, the `U` will be created by
+   * calling `onNone`. If this `Optional` does contain a value, the `U` will be
+   * created by calling `onSome`.
+   *
+   * For the FP enthusiasts in the room, this function:
+   * 1. Could be used to implement all of the functions below
+   * 2. Forms a catamorphism
+   */
+  public fold<U>(onNone: () => U, onSome: (value: T) => U): U {
+    if (this.tag) {
+      return onSome(this.value as T);
+    } else {
+      return onNone();
+    }
+  }
 
-  readonly getOrNull: () => T | null;
-  readonly getOrUndefined: () => T | undefined;
   /**
-  - if some: return self
-  - if none: return opt
-  */
-  readonly or: <T2 = T>(opt: Optional<T2>) => Optional<T | T2>;
+   * Determine if this `Optional` object contains a value.
+   */
+  public isSome(): boolean {
+    return this.tag;
+  }
 
-  /** Same as "or", but uses a thunk instead of a value */
-  readonly orThunk: <T2 = T>(makeOption: () => Optional<T2>) => Optional<T | T2>;
+  /**
+   * Determine if this `Optional` object **does not** contain a value.
+   */
+  public isNone(): boolean {
+    return !this.tag;
+  }
 
-  /** Run a function over the 'some' value.
-   *  "map" operation on the Optional functor.
+  // --- Functor (name stolen from Haskell / maths) ---
+
+  /**
+   * Perform a transform on an `Optional` object, **if** there is a value. If
+   * you provide a function to turn a T into a U, this is the function you use
+   * to turn an `Optional<T>` into an `Optional<U>`. If this **does** contain
+   * a value then the output will also contain a value (that value being the
+   * output of `mapper(this.value)`), and if this **does not** contain a value
+   * then neither will the output.
    */
-  readonly map: <U> (mapper: (x: T) => U) => Optional<U>;
+  public map<U>(mapper: (value: T) => U): Optional<U> {
+    if (this.tag) {
+      return Optional.some(mapper(this.value as T));
+    } else {
+      return Optional.none();
+    }
+  }
 
-  /** Run a side effect over the 'some' value */
-  readonly each: (worker: (x: T) => void) => void;
+  // --- Monad (name stolen from Haskell / maths) ---
 
-  /** "bind"/"flatMap" operation on the Optional Bind/Monad.
-   *  Equivalent to >>= in Haskell/PureScript; flatMap in Scala.
+  /**
+   * Perform a transform on an `Optional` object, **if** there is a value.
+   * Unlike the map function earlier in the piece, here the transform itself
+   * also returns an `Optional`.
    */
```

Comment:
This needs rewording, as you can't say "earlier in the piece" as when viewing the doc string helper for the function it won't have that context and so each comment here needs to be largely standalone.

- [ ] Reword

Reply:
Reworded